### PR TITLE
8316533: C2 compilation fails with assert(verify(phase)) failed: missing Value() optimization

### PR DESCRIPTION
**Pull request by eme64**
**Problem**
We have a `abstract` class `A` with no subtype. Hence, a reference of type `A` must always be `null` (unless a subclass were to be loaded, which we guard against with a compile dependency).

But there are at least these two ways a `A:NotNull` can be created:
- Null-Check: CastPP after null-check improves type from `A` to `A:NotNull`.
- Forced compilation (eg CTW) of a member method of `A`. Then `Parm0` has type `A`, which is improved to `A:NotNull` because the `this/self` pointer cannot be `null`.

This means we are now left with an impossible type `A:NotNull`, a path that uses this type will never be executed.

The question is now what should happen at a `SubTypeCheck` if we do:
`SubTypeCheck(  oop #A:NotNull   ,   constant-classptr-A-exact )`

The verification happens because we do these two different things:
- `SubTypeCheck`: we first detect that we have a constant classptr of a class `A`, which is abstract and has no subtype. Hence, we conclude that any oop compared to it cannot be a subtype (there are no subtypes), and it cannot be of the same type (class is abstract). Hence, any oop must be a supertype (TypeInt::CC_GT).
- The verification code computes the subtype check by computing the klass of the oop via `LoadKlass` (this constant folds to `constant-classptr-A-exact`, because the type of the oop is `A:NotNull`). The `CmpP` node compares the two klasses, and sees that they are identical, and returns an `TypeInt::CC_EQ`.

**Alternatives**

Both results are reasonable, but they are in fact both supersets of the true result. We should take the intersection of the two and get `Type:TOP`, since the input type is already impossible. In fact, it would be best if the impossible type was never created. We could do that by improving `CmpP` to detect the impossible type and constant fold towards the `null` path, removing the `A:NotNull` path. It is harder to deal with the forced-compilation of non-static methods of an abstract class with no subclasses - here we would basically have to forbid compilation or replace the compilation with a `Halt`.

**Solution**

Instead, I have now decided to change the logic in `SubTypeCheckNode` to return `EQ` in case the oop has the same klass and is `NotNull`.

**Testing**

Tier1-6 and stress testing. Running...

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316533](https://bugs.openjdk.org/browse/JDK-8316533): C2 compilation fails with assert(verify(phase)) failed: missing Value() optimization (**Bug** - P4)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16361/head:pull/16361` \
`$ git checkout pull/16361`

Update a local copy of the PR: \
`$ git checkout pull/16361` \
`$ git pull https://git.openjdk.org/jdk.git pull/16361/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16361`

View PR using the GUI difftool: \
`$ git pr show -t 16361`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16361.diff">https://git.openjdk.org/jdk/pull/16361.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16361#issuecomment-1790355141)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * @requires vm.compiler2.enabled
+ * @run main/othervm -XX:CompileCommand=printcompilation,*A::test
+ *                   -XX:CompileCommand=compileonly,*A::test
+ *                   -Xcomp -XX:+IgnoreUnrecognizedVMOptions -XX:+StressReflectiveCode
+ *                   compiler.types.TestSubTypeOfAsbtractClass
+ */
+
+package compiler.types;
+
+public class TestSubTypeOfAsbtractClass {
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java:46]: Typo:
```suggestion
public class TestSubTypeOfAbstractClass {
```

File: test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @summary Oop of abstract class A with no subclass is subtype checked after null-check
+ * @run driver compiler.types.TestSubTypeOfAsbtractClass
+ */
+
+/**
+ * @test
+ * @bug 8316533
+ * @summary Oop of abstract class A is subtype checked after null-check
+ * @requires vm.compiler2.enabled
+ * @run main/othervm -XX:CompileCommand=printcompilation,*A::test
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java:38]: `-XX:CompileCommand=printcompilation,*A::test` is unnecessary.

File: test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    public static void main(String[] args) {
+        for (int i = 0; i < 10_000; i++ ) {
+            A.test();
+        }
+    }
+}
+
+
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java:67]: Please remove these extra line breaks.

File: src/hotspot/share/opto/subtypenode.cpp [hunk 4]
```diff
@@ -43,7 +43,21 @@ const Type* SubTypeCheckNode::sub(const Type* sub_t, const Type* super_t) const
     if (!superklass->is_interface() && superklass->is_abstract() &&
         !superklass->as_instance_klass()->has_subklass()) {
       Compile::current()->dependencies()->assert_leaf_type(superklass);
-      return TypeInt::CC_GT;
+      if (subk->is_same_java_type_as(superk) && !sub_t->maybe_null()) {
+        // The super_t has no subclasses, and sub_t has the same type and is not null,
+        // hence the check should always evaluate to EQ. However, this is an impossible
+        // situation since super_t is also abstract, and hence sub_t cannot have the
+        // same type and be non-null.
+        // Still, if the non-static method of a abstract class without subclasses is
```
Comment (robcasloz) [inline src/hotspot/share/opto/subtypenode.cpp:51]:
```suggestion
        // Still, if the non-static method of an abstract class without subclasses is
```

File: src/hotspot/share/opto/subtypenode.cpp [hunk 5]
```diff
@@ -43,7 +43,21 @@ const Type* SubTypeCheckNode::sub(const Type* sub_t, const Type* super_t) const
...
+        // force-compiled, the Param0 has the self/this pointer with NotNull. This
+        // method would now never be called, because of the leaf-type dependency. Hence,
+        // just for consistency with verification, we return EQ.
+        return TypeInt::CC_EQ;
+      } else {
+        // subk is either a supertype of superk, or null. In either case, superk is a
+        // subtype.
+        return TypeInt::CC_GT;
+      }
```
Comment (robcasloz) [inline src/hotspot/share/opto/subtypenode.cpp:60]: Style suggestion:
```suggestion
      }
      // subk is either a supertype of superk, or null. In either case, superk is a subtype.
      return TypeInt::CC_GT;
```

**Review by robcasloz — APPROVED** [review]
Looks good, I just have a few minor comments and style suggestions.

File: test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java (hunk 1 above)
Comment (eme64) [reply test/hotspot/jtreg/compiler/types/TestSubTypeOfAsbtractClass.java:46]: I had copied the mistake from the other test: `TestSubTypeOfAsbtractClassWrongResult.java`

**Review by robcasloz — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. @rwestrel should take a look as well.

**Review by rwestrel — APPROVED** [review]
Looks reasonable to me.

**Comment by eme64** [comment]
Thanks @TobiHartmann  @robcasloz @rwestrel for the reviews!
/integrate

**Pull request closed without merging**